metricflow/test/snapshots/test_derived_metric_rendering.py/SqlQueryPlan/DuckDB/test_offset_to_grain_metric_filter_and_query_have_different_granularities__plan0_optimized.sql

```diff
@@ -0,0 +1,36 @@
+-- Compute Metrics via Expressions
+SELECT
+  metric_time__month
+  , bookings_start_of_month AS bookings_at_start_of_month
+FROM (
+  -- Constrain Output with WHERE
+  -- Pass Only Elements: ['bookings', 'metric_time__month']
+  -- Aggregate Measures
+  -- Compute Metrics via Expressions
+  SELECT
+    metric_time__month
+    , SUM(bookings) AS bookings_start_of_month
+  FROM (
+    -- Join to Time Spine Dataset
+    -- Pass Only Elements: ['bookings', 'metric_time__month', 'metric_time__day']
+    SELECT
+      subq_13.ds AS metric_time__day
+      , DATE_TRUNC('month', subq_13.ds) AS metric_time__month
+      , subq_11.bookings AS bookings
+    FROM ***************************.mf_time_spine subq_13
+    INNER JOIN (
+      -- Read Elements From Semantic Model 'bookings_source'
+      -- Metric Time Dimension 'ds'
+      SELECT
+        DATE_TRUNC('day', ds) AS metric_time__day
+        , 1 AS bookings
+      FROM ***************************.fct_bookings bookings_source_src_28000
+    ) subq_11
+    ON
+      DATE_TRUNC('month', subq_13.ds) = subq_11.metric_time__day
+    WHERE DATE_TRUNC('month', subq_13.ds) = subq_13.ds
+  ) subq_15
+  WHERE metric_time__day = '2020-01-01'
```

Comment:
This all feels very convoluted but maybe that's just how we have to do offset to grain calculations? Like if this grouped by metric_time__day I guess we'd drop the where filter and just have a repeated bookings_start_of_month value collected alongside each day.

I'm actually not sure how I'd write the SQL for this to make it less confusing. For this case I'd probably do the filter + aggregation on the month grains and then join to time spine to fill out other granularities, but that doesn't generalize.

Anyway, something for us to think about later - it really feels like there has to be a cleaner way to render this, but maybe there isn't and that's another reason why MetricFlow is useful.

Reply:
`offset_to_grain` ends up with the weirdest edge cases. We might just want to be more restrictive with what we allow for those metrics. It's already so weird when there are multiple instances of `metric_time` queried with an `offset_to_grain` metric because the where filter that we add by default doesn't totally do its job anymore. And I have no idea why someone would want to make that query. But alas, for now, we can just give the people what they asked for.

Reply:
> But alas, for now, we can just give the people what they asked for.

LOL, this is me every time I have to add a feature of any kind.